src/main/scala/services/mssql/MsSqlDataProvider.scala

```diff
@@ -4,33 +4,29 @@ package services.mssql
 import models.schemas.DataRow
 import services.mssql.MsSqlConnection.VersionedBatch
 import services.mssql.base.MssqlVersionedDataProvider
-import services.streaming.base.HasVersion
-
+import com.sneaksanddata.arcane.framework.logging.ZIOLogAnnotations.zlog
 import zio.stream.ZStream
 import zio.{Task, ZIO, ZLayer}
 
 import java.time.Duration
 import scala.util.{Failure, Try}
 
-given HasVersion[VersionedBatch] with
-  type VersionType = Option[Long]
 
-  private val partial: PartialFunction[VersionedBatch, Option[Long]] =
+extension (result: VersionedBatch)
+  def getLatestVersion: ZIO[Any, Throwable, Option[Long]] = result match
     case (queryResult, version: Long) =>
-      // If the database response is empty, we can't extract the version from it and return the old version.
-      queryResult.read.headOption match
-        case None      => Some(version)
+      queryResult.read.runHead.flatMap {
+        // If the database response is empty, we can't extract the version from it and return the old version.
+        case None => ZIO.succeed(Some(version))
         case Some(row) =>
           // If the database response is not empty, we can extract the version from any row of the response.
           // Let's take the first row and try to extract the version from it.
           val dataVersion = row.filter(_.name == "ChangeTrackingVersion") match
-            // For logging purposes will be used in the future.
-            case Nil          => Failure(new UnsupportedOperationException("No ChangeTrackingVersion found in row."))
-            case version :: _ => Try(version.value.asInstanceOf[Long])
-          dataVersion.toOption
-
-  extension (result: VersionedBatch)
-    def getLatestVersion: this.VersionType = partial.applyOrElse(result, (_: VersionedBatch) => None)
+            // Hard fail when ChangeTrackingVersion is not found
+            case Nil => ZIO.die(new Throwable("No ChangeTrackingVersion found in row."))
+            case version :: _ => ZIO.attempt(version.value.asInstanceOf[Long]).map(Option.apply).orElse(ZIO.succeed(Option.empty[Long]))
```

Comment:
```suggestion
            case version :: _ => ZIO.attempt(version.value.asInstanceOf[Long]).map(Option.apply).orElse(ZIO.succeed(None))
```

`None` is better here because the type can be inferred.